Asynchronous processing

htlyu opened the pull request:
Convert file processing, entity extraction, and similar operations to asynchronous execution.

htlyu committed:
Convert file processing, entity extraction, and similar operations to asynchronous execution.

htlyu committed:
format the document

MaKaili commented on src/hirag_mcp/hirag.py:
```diff
@@ -93,9 +96,28 @@ async def insert_to_kb(
+        loader_configs: Optional[dict] = None,
+    ):
+        start_total = time.perf_counter()
+        documents = load_document(
+            document_path, content_type, document_meta, loader_configs
+        )
```
Should change to a async function, otherwise there is risk to block all coroutines when there is a large file. Considering to use `ThreadPoolExecutor`

MaKaili commented on src/hirag_mcp/hirag.py:
```diff
@@ -93,9 +96,28 @@ async def insert_to_kb(
-            # TODO: handle the concurrent upsertion
-            for relation in relations:
-                await self.gdb.upsert_relation(relation)
+        await asyncio.gather(*tasks)
```
Consider to set the limit to the concurrent number for each asyncio.gather by using `asyncio.Semaphore`

MaKaili commented on src/hirag_mcp/hirag.py:
```diff
@@ -50,27 +61,14 @@ async def create(cls, **kwargs):
-                await self.vdb.upsert_text(
+    async def _process_document(self, document):
+        loop = asyncio.get_event_loop()
+        with ThreadPoolExecutor() as pool:
```
`chunk()` is a computation bound instead of io bound, consider to change to a proper concurrent method

htlyu committed:
feat(rag): convert file loading to async; add semaphore for concurrency control; switch chunk processing to multithreading

Use asyncio.to_thread for load_document
Introduce asyncio.Semaphore to throttle concurrent upserts
Run chunker.chunk() in a ProcessPoolExecutor for multithreading

htlyu committed:
Limit the concurrency of chunks, entities, and relations separately
